experimental/dds/tree2/src/feature-libraries/editable-tree-2/api/node.ts

```diff
@@ -0,0 +1,81 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { assert } from "@fluidframework/core-utils";
+import { fail } from "../../../util";
+import {
+	FieldSchema,
+	schemaIsFieldNode,
+	schemaIsLeaf,
+	schemaIsMap,
+	schemaIsStruct,
+} from "../../typed-schema";
+import { FieldKinds } from "../../default-field-kinds";
+import { EmptyKey } from "../../../core";
+import { TreeNode, TypedField } from "../editableTreeTypes";
+import { createListProxy } from "./list";
+import { createObjectProxy } from "./object";
+
+/** Symbol used to store a private/internal reference to the underlying editable tree node. */
+const treeNodeSym = Symbol("TreeNode");
+
+/** Helper to retrieve the stored tree node. */
+export function getTreeNode(target: object): TreeNode {
+	return (target as any)[treeNodeSym] as TreeNode;
+}
+
+/** Helper to set the stored tree node. */
+export function setTreeNode(target: any, treeNode: TreeNode) {
+	Object.defineProperty(target, treeNodeSym, {
+		value: treeNode,
+		writable: false,
+		enumerable: false,
+		configurable: false,
+	});
+}
+
+// TODO: Implement lifetime.  Referential identity is tied to the proxy instance (not the target, etc.)
+// Therefore, it is the proxy that should be cached on their respective nodes and reused.
+
+/** Retrieve the associated proxy for the given field. */
+export function getProxyForField<T extends FieldSchema>(field: TypedField<T>) {
+	switch (field.schema.kind) {
+		case FieldKinds.required: {
+			const asValue = field as TypedField<FieldSchema<typeof FieldKinds.required>>;
+
+			// TODO: The only reason for using 'boxedContent' is to prevent FieldNode/LazySequence from
+			//       collapsing to 'undefined'.  Otherwise, we could avoid the overhead of allocating
```

Comment:
When does this collapsing happpen? In `getProxyForNode`, I would expect the call to `getField` in the "Field node" case to always return a TreeField, never `undefined`. Is that not the case?